The RF Index tool is designed to automatically generate a Bowtie reference index, that will be used by the RF Count module for reads mapping.<br />This tool requires an internet connection, since it relies on querying the UCSC Genome database to obtain transcripts annotation and reference genome’s sequence.<br /><br />

# Usage
To list the required parameters, simply type:

```bash
$ rf-index -h
```

Parameter         | Type | Description
----------------: | :--: |:------------
__-o__ *or* __--output-dir__ | string | Bowtie index output directory (Default: &lt;assembly&gt;\_&lt;annotation&gt;, e.g. “mm9_refFlat/”)
__-ow__ *or* __--overwrite__ | | Overwrites the output directory if already exists
__-g__ *or* __--genome-assembly__ | string |Genome assembly for the species of interest (Default: __mm9__).<br /> For a complete list of UCSC available assemblies, please refer to the UCSC website (<https://genome.ucsc.edu/FAQ/FAQreleases.html>)
__-a__ *or* __--annotation__ | string | Name of the UCSC table containing the genes annotation (Default: __refFlat__).<br />For a complete list of tables available for the chosen assembly, please refer to the UCSC website (<https://genome.ucsc.edu/cgi-bin/hgTables>)
__-n__ *or* __--gene-name__ | | When possible, gene names will be used instead of gene IDs/accessions
__-t__ *or* __--timeout__ | int | Connection’s timeout in seconds (Default: __180__)
__-r__ *or* __--reference__ | string | Path to a FASTA file containing chromosome (or scaffold) sequences for the chosen genome assembly.<br/>__Note:__ if no file is specified, RF Index will try to obtain sequences from the UCSC DAS server. This process may take up to hours, depending on your connection's speed.
__-b__ *or* __--bowtie-build__ | string | Path to ``bowtie-build`` executable (Default: assumes bowtie-build is in PATH)
__-e__ *or* __--bedtools__ | string | Path to ``bedtools`` executable (Default: assumes ``bedtools`` is in PATH)

!!! note "Note"
    For RNA structure probing experiments conducted over synthetic RNAs (or custom RNA pools), a reference can be generated by directly invoking the ``bowtie-build`` command, however it is first necessary to sort lexicographically the FASTA file by sequence ID.

To prepare a custom Bowtie index, simply do:

```bash
# Sort reference FASTA file by sequence ID
$ awk 'BEGIN{RS=">"} NR>1 {gsub("\n", "\t"); print ">"$0}' reference_unsorted.fa | \ 
  LC_ALL=C sort -T "" -t ' ' -k2,2 | \ 
  awk '{sub("\t", "\n"); gsub("\t", ""); print $0}' > reference_sorted.fa

# Build the Bowtie index
$ bowtie-build reference_sorted.fa reference_sorted

$ ls -l

  -rwxrwxrwx 1 danny epigenetics  96041105 5 mar 10.50 reference_sorted.1.ebwt  -rwxrwxrwx 1 danny epigenetics  37313744 5 mar 10.50 reference_sorted.2.ebwt  -rwxrwxrwx 1 danny epigenetics   1844468 5 mar 10.28 reference_sorted.3.ebwt  -rwxrwxrwx 1 danny epigenetics  74627475 5 mar 10.28 reference_sorted.4.ebwt  -rwxrwxrwx 1 danny epigenetics 302198817 5 mar 10.28 reference_sorted.fa  -rwxrwxrwx 1 danny epigenetics  96041105 5 mar 11.11 reference_sorted.rev.1.ebwt  -rwxrwxrwx 1 danny epigenetics  37313744 5 mar 11.11 reference_sorted.rev.2.ebwt  -rwxrwxrwx 1 danny epigenetics 302198817 5 mar 10.28 reference_unsorted.fa
```