The RF Norm module takes one (Rouskin and Zubradt methods), two (Ding method), or three (Siegfried method) RC files generated by the RF Count module, and performs normalization to obtain transcriptome-wide per-base reactivities.<br/>Reactivity scores can be computed using 4 methods:
<br/><br/>
## Scoring of RT-stops/nuclease cuts-based methods
__[1] Ding *et al*., 2014__<br/><br/>
Per-base signal is calculated as the natural log (ln) of the ratio between the raw count of RT-stops/nuclease cuts at a given position of a transcript, and the average of the ln of RT-stops/nuclease cuts along the whole transcript:<br/><br/>

<math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><msub><mi>U</mi><mi>i</mi></msub><mo>=</mo><mfrac><mrow><mi>ln</mi><mo>(</mo><msub><mi>n</mi><mrow><mn>1</mn><mi>i</mi></mrow></msub><mo>+</mo><mi>p</mi><mo>)</mo></mrow><mfenced><mrow><munderover><mo>&#x2211;</mo><mrow><mi>j</mi><mo>=</mo><mn>0</mn></mrow><mi>l</mi></munderover><mstyle displaystyle="true"><mfrac><mrow><mi>ln</mi><mo>(</mo><msub><mi>n</mi><mrow><mn>1</mn><mi>j</mi></mrow></msub><mo>+</mo><mi>p</mi><mo>)</mo></mrow><mi>l</mi></mfrac></mstyle></mrow></mfenced></mfrac></math>
<br/>
<math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><msub><mi>T</mi><mi>i</mi></msub><mo>=</mo><mfrac><mrow><mi>ln</mi><mo>(</mo><msub><mi>n</mi><mrow><mn>2</mn><mi>i</mi></mrow></msub><mo>+</mo><mi>p</mi><mo>)</mo></mrow><mfenced><mrow><munderover><mo>&#x2211;</mo><mrow><mi>j</mi><mo>=</mo><mn>0</mn></mrow><mi>l</mi></munderover><mstyle displaystyle="true"><mfrac><mrow><mi>ln</mi><mo>(</mo><msub><mi>n</mi><mrow><mn>2</mn><mi>j</mi></mrow></msub><mo>+</mo><mi>p</mi><mo>)</mo></mrow><mi>l</mi></mfrac></mstyle></mrow></mfenced></mfrac></math>
<br/>
where *n<sub>1i</sub>* and *n<sub>2i</sub>* are respectively the raw read counts in the untreated (or RNase V1) and treated (DMS, CMCT, SHAPE, or Nuclease S1) samples at position *i* of the transcript, *l* is the transcriptâ€™s length, and *p* is a pseudocount added to deal with non-covered regions. *U<sub>i</sub>* and *T<sub>i</sub>* are respectively the normalized number of RT-stops/nuclease cuts at position *i* in the untreated and treated samples.<br/>Score at position i is then calculated as:<br/><br/>
<math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><msub><mi>S</mi><mi>i</mi></msub><mo>=</mo><mi>m</mi><mi>a</mi><mi>x</mi><mo>(</mo><mn>0</mn><mo>,</mo><mo>&#xA0;</mo><mo>(</mo><msub><mi>T</mi><mi>i</mi></msub><mo>-</mo><msub><mi>U</mi><mi>i</mi></msub><mo>)</mo><mo>)</mo></math>
<br/>
__[2] Rouskin *et al*., 2014__<br/><br/>
The untreated sample is not considered. Per-base RT-stops/nuclease cuts are used as a direct measure of the raw signal.

!!! warning "Warning"
    Normalization of data processed by Rouskin method, can only be performed using the __90% Winsorising__ approach.
<br/>
## Scoring of mutational profiling-based methods
__[3] Siegfried *et al*., 2014__<br/><br/>
This method takes into account both an untreated sample, and a denatured controlsample.<br/>Per-base raw signal is calculated as:<br/><br/>

<math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><msub><mi>S</mi><mi>i</mi></msub><mo>=</mo><mfrac><mrow><mstyle displaystyle="true"><mfrac><msub><mi>n</mi><mrow><mi>T</mi><mi>i</mi></mrow></msub><msub><mi>c</mi><mrow><mi>T</mi><mi>i</mi></mrow></msub></mfrac></mstyle><mo>-</mo><mstyle displaystyle="true"><mfrac><msub><mi>n</mi><mrow><mi>U</mi><mi>i</mi></mrow></msub><msub><mi>c</mi><mrow><mi>U</mi><mi>i</mi></mrow></msub></mfrac></mstyle></mrow><mstyle displaystyle="true"><mfrac><msub><mi>n</mi><mrow><mi>D</mi><mi>i</mi></mrow></msub><msub><mi>c</mi><mrow><mi>D</mi><mi>i</mi></mrow></msub></mfrac></mstyle></mfrac></math>
<br/>
where *n<sub>Ti</sub>*, *n<sub>Ui</sub>*, *n<sub>Di</sub>* are respectively the mutation counts in the treated, untreated, and denatured samples at position *i* of the transcript, while *c<sub>Ti</sub>*, *c<sub>Ui</sub>*, *c<sub>Di</sub>* are respectively the reads covering position *i* of the transcript in the treated, untreated, and denatured samples.
<br/><br/>
__[4] Zubradt *et al*., 2016__<br/><br/>
The untreated sample is not considered. Per-base raw signal is calculated as:<br/><br/>
<math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><mfrac><msub><mi>n</mi><mrow><mi>T</mi><mi>i</mi></mrow></msub><msub><mi>c</mi><mrow><mi>T</mi><mi>i</mi></mrow></msub></mfrac></math>
<br/>
where *n<sub>Ti</sub>*, and *c<sub>Ti</sub>* are respectively the mutations count and the reads coverage at position *i* of the transcript.
<br/>
## Normalization of raw reactivities
Raw reactivity scores can be normalized using 3 different approaches:<br/><br/>

Method         | Description
-------------: | :------------
__2-8% Normalization__ | From the top 10% of values, the top 2% is ignored, then any reactivity value along the entire transcript is divided by the average of the remaining 8%
__90% Winsorising__ | Each reactivity value above the 95<sup>th</sup> percentile is set to the 95<sup>th</sup> percentile, and the reactivity at each position of the transcript is divided by the value of the 95<sup>th</sup> percentile
__Box-plot Normalization__ | Values greater than 1.5x the interquartile range (numerical distance between the 25<sup>th</sup> and 75<sup>th</sup> percentiles) above the 75<sup>th</sup> percentile are removed. After excluding these outliers, the next 10% of reactivities are averaged, and all reactivities (including outliers) are divided by this value.
<br/>
Normalized reactivities can be further remapped to values ranging from 0 to 1 according to Zarringhalam *et al.*, 2012. In this approach, values < 0.25 are linearly mapped to [0-0.35[, values &ge; 0.25 and < 0.3 are linearly mapped to [0.35-0.55[, values &ge; 0.3 and < 0.7 are linearly mapped to [0.55-0.85[, and values &ge; 0.7 are linearly mapped to [0.85-1].<br /><br />

# Usage
To list the required parameters, simply type:

```bash
$ rf-norm -h
```

Parameter         | Type | Description
----------------: | :--: |:------------
__-p__ *or* __--processors__ | int | Number of processors (threads) to use (Default: __1__)